flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresStreamFetchTask.java

```diff
@@ -128,6 +128,18 @@ public void commitCurrentOffset() {
                             postgresOffsetContext
                                     .getOffset()
                                     .get(PostgresOffsetContext.LAST_COMMIT_LSN_KEY);
+
+            if (split != null && split.isStreamSplit()) {
+                // Note: We should commit the LSN from the splits in the checkpoint, not the
+                // context's LSN. This is because after the source completes the snapshot function,
+                // it will begin consuming data from the PostgreSQL replication stream.
+                // Within PostgresStreamingChangeEventSource, the context's LSN will be updated.
+                commitLsn =
+                        ((PostgresOffset) split.asStreamSplit().getStartingOffset())
```

Comment:
The biggest problem of this PR: each split is read between [start, end). If commit starting offset  and then failover. The message of starting offset will lose.

Reply:
Perhaps what you're talking about is the start and end of the primary key?
This PR is an attempt to solve the data lost problem in the stream stage, where the root cause is that the offset of the streamsplit in the checkpoint state lags behind the confirmed_flush_lsn of the postgres slot, the wal data between them has been lost, such as the reproducing test case added by the first commit of this PR.

There is no problem in scan/snapshot stage.